chalice/deploy/packager.py

```diff
@@ -1,60 +1,72 @@
 from __future__ import print_function
+import sys
 import hashlib
 import inspect
 import os
 import shutil
 import subprocess
 import zipfile
+import tempfile
+import contextlib
+import pip
 
 import virtualenv
-from typing import Any, Optional  # noqa
+from typing import Any, List, Optional, Tuple, Iterable, Callable  # noqa
+from chalice.compat import StringIO
+from chalice.utils import OSUtils
 
 import chalice
 from chalice import compat, app
 
 
+class InvalidSourceDistributionNameError(Exception):
+    pass
+
+
 class LambdaDeploymentPackager(object):
     _CHALICE_LIB_DIR = 'chalicelib'
     _VENDOR_DIR = 'vendor'
 
-    def _create_virtualenv(self, venv_dir):
-        # type: (str) -> None
-        virtualenv.create_environment(venv_dir)
+    def __init__(self, osutils=None):
+        # type: (Optional[OSUtils]) -> None
+        if osutils is None:
+            osutils = OSUtils()
+        self._osutils = osutils
 
-    def create_deployment_package(self, project_dir, package_filename=None):
+    def _get_requirements_file(self, project_dir):
+        # type: (str) -> str
+        # Gets the path to a requirements.txt file out of a project dir path
+        return os.path.join(project_dir, 'requirements.txt')
+
+    def create_deployment_package(self, project_dir, package_filename=None,
+                                  dependency_builder=None):
         # type: (str, Optional[str]) -> str
         print("Creating deployment package.")
-        # pip install -t doesn't work out of the box with homebrew and
-        # python, so we're using virtualenvs instead which works in
-        # more cases.
-        venv_dir = os.path.join(project_dir, '.chalice', 'venv')
-        self._create_virtualenv(venv_dir)
-        pip_exe = compat.pip_script_in_venv(venv_dir)
-        assert os.path.isfile(pip_exe)
-        # Next install any requirements specified by the app.
-        requirements_file = os.path.join(project_dir, 'requirements.txt')
+        # Now we need to create a zip file and add in the site-packages
+        # dir first, followed by the app_dir contents next.
+        if dependency_builder is None:
+            dependency_builder = DependencyBuilder(OSUtils())
         deployment_package_filename = self.deployment_package_filename(
             project_dir)
         if package_filename is None:
             package_filename = deployment_package_filename
-        if self._has_at_least_one_package(requirements_file) and not \
-                os.path.isfile(package_filename):
-            p = subprocess.Popen([pip_exe, 'install', '-r', requirements_file],
-                                 stdout=subprocess.PIPE)
-            p.communicate()
-        deps_dir = compat.site_packages_dir_in_venv(venv_dir)
-        assert os.path.isdir(deps_dir)
-        # Now we need to create a zip file and add in the site-packages
-        # dir first, followed by the app_dir contents next.
+        if not os.path.isfile(package_filename):
```

Comment:
Probably would want this built into osutils as a `file_exists()` to be consistent on using the OSUtils abstraction